lib/terratag-action.spec.ts

```diff
@@ -0,0 +1,73 @@
+jest.mock("axios");
+jest.mock("child_process");
+import axios from "axios";
+const mockedAxios = axios as jest.Mocked<typeof axios>;
+import childProcess from "child_process";
+const mockedChildProcess = childProcess as jest.Mocked<typeof childProcess>;
+import core from "@actions/core";
+const mockedCore = core as jest.Mocked<typeof core>;
+import tc from "@actions/tool-cache";
+const mockedTC = tc as jest.Mocked<typeof tc>;
+
+import run from "./terratag-action";
+
+test("simple end to end, latest terratag", async () => {
+  jest.resetAllMocks();
+  mockedCore.getInput.mockImplementation((name: string) => {
+    const inputs: { [key: string]: string } = { terratagVersion: "latest", tags: JSON.stringify({ a: "b" }) };
+    return inputs[name];
+  });
+  mockedAxios.get.mockResolvedValue({
+    status: 200,
+    data: `href="/env0/terratag/releases/tag/v1.2.3"`,
+  });
+  mockedTC.downloadTool.mockResolvedValue("FAKE PATH FOR DOWNLOADED TOOL TAR");
+  mockedTC.extractTar.mockResolvedValue("FAKE PATH FOR EXTRACTED");
+  const spawn = {
+    stdout: { on: jest.fn() },
+    stderr: { on: jest.fn() },
+    on: jest.fn(),
+  };
+  spawn.on.mockImplementation((eventName: string, callback: (code: number) => void) => {
+    expect(eventName).toBe("close");
+    callback(0);
+  });
+  mockedChildProcess.spawn.mockReturnValue(spawn as any);
+  await run();
+  expect(mockedCore.addPath.mock.calls).toEqual([["FAKE PATH FOR EXTRACTED"]]);
+  expect(mockedAxios.get.mock.calls).toEqual([["https://github.com/env0/terratag/releases"]]);
+  expect(mockedTC.downloadTool.mock.calls).toEqual([
+    ["https://github.com/env0/terratag/releases/download/v1.2.3/terratag_1.2.3_linux_amd64.tar.gz"],
+  ]);
+  expect(mockedTC.extractTar.mock.calls).toEqual([["FAKE PATH FOR DOWNLOADED TOOL TAR"]]);
+  expect(mockedChildProcess.spawn.mock.calls).toEqual([["FAKE PATH FOR EXTRACTED/terratag", ['-tags={"a":"b"}']]]);
```

Comment:
1. Let's use the more commonly accepted bdd jargon of `it('should do this')`
2. It's better to have one `expect` per `it` - so that if one fail the rest continue to run - you should separate the the "given"/"when" part from the "then".  

Something like:
```typescript
describe('terratag action', () => {
  describe('on success', () => {
    beforeEach(() => { /* mock a happy path and run() */ })
    describe('downloading terratag', () => {
          it('should download the latest terratag version', () => {/* whatever you expect */});
          it('should extract the downloaded tar', () => {/* whatever you expect */});
    });

    it('should spawn a process invoking terratag', () => {});

    describe('supported flags', () => {
      it('should set the dir flag', () => {});
      /* ... */
    });
  });
});
```

Same for errors.

Reply:
lost you here. sorry. must be missing something:
1. if mocks are reset between every it, how can i expect in a different `it` than where it run?
2. if i have different mock setup between cases, how can the `run` be executed in the `beforeEach`?

Reply:
You use `beforeEach` on different levels to set different mocks on different `describe` levels.  
It's alright to call `run()` several times after having mocked it differently of course, but having each `expect()` in a separate `it()`, you'll have shared mocks and run calls for sure.